### ISSUE TITLE
[Bug] bfloat8_b reciprocal/rsqrt/log/div return all-zeros on sub-tile 4-D tensors (silent corruption)

### ISSUE THREAD

**Issue by nkapreTT** [tenstorrent/tt-metal#51831]
## Summary

`ttnn.reciprocal`, `ttnn.rsqrt`, `ttnn.log`, `ttnn.log2`, and `ttnn.log10` on a
**`bfloat8_b`, `TILE_LAYOUT`, 4-D tensor whose last two dims are each smaller than
the 32×32 tile** return an **all-zero output**. `reciprocal` of a strictly
positive input can never be 0, so this is silent data corruption, not a precision
shortfall. `ttnn.div` (which is reciprocal-based) is affected more broadly — it is
wrong even on a plain tile-aligned 2-D tensor.

The same ops are correct in `bfloat16`/`float32`, and the `from_torch → to_torch`
identity roundtrip is correct for `bfloat8_b` at these shapes, so the defect is in
the block-float unary/reciprocal compute path for sub-tile 4-D tensors, not in
tilization.

## Impact

`rsqrt` is the core primitive of **RMSNorm** (Llama, Qwen, Mistral, Gemma, …). A
model running RMSNorm in `bfloat8_b` over an activation with small trailing
spatial dims would silently zero its normalization output — no NaN, no assert.

## Environment

- Arch: **Blackhole p100a**
- tt-metal: `main` @ `d69fab9610d`
- Firmware bundle: 19.9.0

## Minimal repro

```python
import torch, ttnn
from tests.ttnn.utils_for_testing import check_with_pcc

d = ttnn.open_device(device_id=0)
for dtype, name in ((ttnn.bfloat8_b, "bf8_b"), (ttnn.bfloat16, "bf16")):
    for shape in ([1, 2048, 7, 7], [1, 2048, 32, 32]):
        x = torch.randn(shape).abs() + 0.5            # strictly positive
        a = ttnn.from_torch(x, dtype=dtype, layout=ttnn.TILE_LAYOUT, device=d)
        ident = ttnn.to_torch(a)                       # identity roundtrip
        out = ttnn.to_torch(ttnn.reciprocal(a))        # op under test
        print(name, shape,
              "identity_pcc=", str(check_with_pcc(x, ident, 0.99)[1])[:8],
              "reciprocal_pcc=", str(check_with_pcc(torch.reciprocal(x), out, 0.99)[1])[:8],
              "out_max=", float(out.max()))
ttnn.close_device(d)
```

## Observed (measured on the environment above)

| dtype | shape | identity roundtrip PCC | reciprocal PCC | output max |
| --- | --- | --- | --- | --- |
| bfloat8_b | `[1, 2048, 7, 7]` | 0.999917 | **0.0** | **0.0** |
| bfloat8_b | `[1, 64, 7, 7]` | 0.999916 | **0.0** | **0.0** |
| bfloat8_b | `[1, 2048, 8, 8]` | 0.999913 | **0.0** | **0.0** |
| bfloat8_b | `[1, 2048, 32, 32]` | 0.999893 | 0.9995 | 2.0 |
| bfloat16 | `[1, 2048, 7, 7]` | 0.999992 | 0.999983 | 2.0 |

So: block-float specific (bf16 fine), sub-tile-4-D specific (32×32 face fine), and
not the tilization roundtrip (identity fine).

### Also affected in `bfloat8_b`

| op | shape | PCC |
| --- | --- | --- |
| `rsqrt`, `log`, `log2`, `log10` | `[1, *, 7, 7]` | **0.0** (same as reciprocal) |
| `div` | `[32, 2048]` (tile-aligned) | **0.053** |
| `div` | `[1, 256, 7, 7]` | **0.0** |

`div` is wrong even on an aligned 2-D tensor, consistent with a shared
`bfloat8_b` reciprocal path.

## Expected

`bfloat8_b` `reciprocal`/`rsqrt`/`log*`/`div` should produce correct (if lower
precision) results on 4-D sub-tile tensors, as they do in `bfloat16`, rather than
silently returning zeros.

---

<sub>Found by a systematic op × dtype × shape device sweep. Happy to provide the
full per-cell data.</sub>

**Comment by nkapreTT** [comment]
A wider sweep (64 ops) confirms the blast radius is the whole `bfloat8_b` sqrt/log/reciprocal family plus reductions:

- `acosh` (sqrt/log-based) also returns **0.0** on the sub-tile 4-D shape `[1, 256, 7, 7]`.
- `max`/`min` reductions over the last dim return **0.0** on the same sub-tile 4-D shape.
- `div` degrades with tensor size in bfloat8_b: PCC **0.020** at `[32, 2048]`, **0.002** at `[64, 4096]`, **0.0** at 4-D — consistently wrong even on tile-aligned 2-D.

`reciprocal`, `rsqrt`, `log`, `log2`, `log10`, `acosh`, `div`, and `max`/`min` are all correct in bfloat16/float32 at these shapes.

**Comment by nkapreTT** [comment]
A 121-op systematic sweep (Blackhole p100a, same build) shows this is **two distinct bfloat8_b bugs**:

**Bug A — division is broken on *every* shape (not just sub-tile 4-D):**

| op | bf8_b `[32,2048]` (tile-aligned) | bf8_b `[1,256,7,7]` | bf16/fp32 |
|---|---|---|---|
| `div` | **0.019** | 0.0 | ~1.0 |
| `div_no_nan` | **0.035** | 0.0 | ~1.0 |
| `floor_div` | **0.047** | 0.0 | ~1.0 |

**Bug B — sub-tile 4-D returns all-zeros** (the originally reported case), now confirmed across the reciprocal/log/special-function family: `reciprocal`, `rsqrt`, `log`, `log2`, `log10`, `logit`, `xlogy`, `acosh`, `digamma`, `polygamma`, `remainder`, `fmod`, `rdiv`, and `max`/`min` reductions — all ~0.99 on tile-aligned 2-D but **0.0** on `[1,256,7,7]` in bfloat8_b, while correct in bfloat16/float32.

Both point at the shared bfloat8_b reciprocal/division path. Bug A is the higher-impact one — plain `ttnn.div` in bfloat8_b is unusable regardless of shape.

**Comment by nkapreTT** [comment]
The bf8_b division/reciprocal corruption also **propagates into the backward (gradient) kernels** — confirmed via an autograd-golden sweep of 82 `*_bw` ops on the same Blackhole p100a:

| backward op | bf16 grad PCC | **bf8_b grad PCC** | fp32 |
|---|---|---|---|
| `div_bw` | 0.99999 | **0.112** | ~1.0 |
| `reciprocal_bw` | 0.99999 | **0.098** | ~1.0 |
| `rdiv_bw` | 0.99999 | **0.101** | ~1.0 |

and the sub-tile-4-D zeros (`rsqrt_bw`/`log_bw`/`sqrt_bw`/… → 0.0 on `[1,256,7,7]` bf8_b). So training in bfloat8_b through any of these ops silently corrupts gradients, not just the forward output.

**Comment by kanapitsas** [comment]
I reproduced the five unary failures on Blackhole ttsim at `a3ece1ae1e` (simulator, not
silicon) and narrowed the trigger to zero padding in the last dimension.

The host packer documents `bfloat8_b` as one shared exponent per 16-element row segment:
[`get_max_exp()`](https://github.com/tenstorrent/tt-metal/blob/9fb0ed54794395ad68a73208e483fddb067a01fe/tt_metal/impl/data_format/blockfloat_common.cpp#L26-L50) requires 16 values, and [`pack_as_bfp_tiles()`](https://github.com/tenstorrent/tt-metal/blob/9fb0ed54794395ad68a73208e483fddb067a01fe/tt_metal/impl/data_format/blockfloat_common.cpp#L409-L447) calls it once per face row.

The failure matches this path: default zero padding reaches `reciprocal`, `rsqrt`, `log`,
`log2`, and `log10`, which produce non-finite values there. Packing the result back to
`bfloat8_b` then makes finite values in the same 16-lane group underflow to zero.

Controls:

- Sampled widths 1, 2, 4, 8, 15 fail completely; 16 is clean; 17, 24, 31 fail only in the
  trailing group; 32 is clean.
- The same sampled heights (1 … 32) are all clean when width is 32.
- `from_torch(..., pad_value=1.0)` removes the failure for all five ops. Max relative error
  then falls to 0.009–0.031 against torch on the quantized input.
- One zero in a full 32×32 tile zeros exactly the other 15 lanes in its group.

The width/height asymmetry is the part that does not fit "partial tiles" or per-face
exponent sharing: padding *rows* form their own 16-lane groups, so their non-finite values
stay contained, while padding *columns* sit inside groups that hold real data.

I could not reproduce a separate shape-only `div` failure. On the exact `[32, 2048]` shape,
with denominators in `[1, 5]`, there were no zero outputs and max relative error was 0.0625;
four aligned 4-D controls gave the same result. Adding one zero denominator zeroed the other
15 lanes in its group. A pack-only control shows the same loss once one value is 128× larger
than its neighbors, so that part is normal block-float dynamic-range loss. The original
aligned `div` result still needs its exact input data before it can be classified as the
same mechanism.

So the confirmed bug is narrower: implementation padding reaches pole-at-zero ops and
corrupts logical data. `pad_value=1.0` is a caller-side workaround for these five ops, not a
general fix.

Could someone confirm the `pad_value=1.0` control on p100a? If it matches, widths 15/16/17
would make a compact regression set.
